core/common/src/main/java/alluxio/conf/InstancedConfiguration.java

```diff
@@ -453,15 +467,29 @@ private String lookupRecursively(String base, Set<String> seen)
       if (!PropertyKey.isValid(match)) {
         throw new RuntimeException(ExceptionMessage.INVALID_CONFIGURATION_KEY.getMessage(match));
       }
-      String value = lookupRecursively(mProperties.get(PropertyKey.fromString(match)), seen);
+      key = PropertyKey.fromString(match);
+      Object value = mProperties.get(key);
+      String stringValue = null;
+      if (value instanceof String) {
+        stringValue = String.valueOf(lookupRecursively((String) value , seen));
+      }
+      else if (value != null) {
+        stringValue = String.valueOf(mProperties.get(key));
+      }
       seen.remove(match);
-      if (value == null) {
+      if (stringValue == null) {
         throw new UnresolvablePropertyException(ExceptionMessage
             .UNDEFINED_CONFIGURATION_KEY.getMessage(match));
       }
-      resolved = resolved.replaceFirst(REGEX_STRING, Matcher.quoteReplacement(value));
+      resolved = resolved.replaceFirst(REGEX_STRING, Matcher.quoteReplacement(stringValue));
+    }
+    if (key != null) {
+      resolvedValue = key.parseValue(resolved);
+    }
+    else {
```

Comment:
nit: no line break before `else` as code convention in this repo.
We basically follow Google Java style guide: https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style

```suggestion
    } else {
```